kubelet/kubenet: split hostport handling into separate module

dcbw opened the pull request:
This pulls the hostport functionality of kubenet out into a separate module so that it can be more easily tested and potentially used from other code (maybe CNI, maybe downstream consumers like OpenShift, etc).  Couldn't find a mock iptables so I wrote one, but I didn't look very hard.

@freehan @thockin @bprashanth

dcbw committed:
kubelet/kubenet: split hostport handling into separate module

dcbw commented:
2016/06/07 07:20:44 e2e.go:212: Running: build-release
ERROR: gcloud crashed (error): [Errno 111] Connection refused

dcbw commented:
@k8s-bot test this issue: #23545

dcbw commented:
Sigh... a couple instances of:

      Timed out after 300.000s.
      Expected
          <*errors.errorString | 0xc820acd9b0>: {
              s: "expected metrics for kubelet",
          }
      to be nil

which seems to be flake #26431

dcbw commented:
@k8s-bot test this issue: #26431

dcbw commented:
cc @danwinship @pravisankar

freehan commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -347,14 +321,20 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
+	// Open any hostports the pod's containers want
+	runningPods, err := plugin.getRunningPods()
+	if err == nil {
+		err = plugin.hostportHandler.OpenPodHostportsAndSync(pod, BridgeName, runningPods)
```
Say if the CNI ADD succeeded, but port is in use. SetUpPod will end up setting up the veth but fail at opening hostport. IIUC, kubelet will just kill infra container without calling TearDownPod (it assumes SetUpPod won't partially succeed). Then this Pod will stuck, because CNI ADD will always return failure since eth0 has been setup.

freehan replied:
Need some restructure. Other parts looks good overall.

dcbw replied:
> Say if the CNI ADD succeeded, but port is in use. SetUpPod will end up setting up the veth but fail at opening hostport. IIUC, kubelet will just kill infra container without calling TearDownPod (it assumes SetUpPod won't be partially successful). Then this Pod will stuck, because CNI ADD will always return failure since eth0 has been setup.

You are correct that this PR will exacerbate the situation a bit, so I'll try to fix that.  However, it looks like all the network plugins have problems with this, since they don't necessarily clean up after themselves on errors in SetUpPod().  And since (at least) dockertools/manager.go never calls TearDownPod() when setup fails, I think there are already cases where failures will leave the pod network configured...  I'll try to figure something out here.

dcbw replied:
I don't think the pod will actually get stuck, since eth0 will get destroyed along with the infra container's network namespace.  But what won't be cleaned up is any IPAM leases or shaper stuff.

freehan replied:
oh that is even worse, because kubelet will keep retry and used up all IPs. :(

dcbw replied:
Researching this more indicates that the docker runtime will eventually tear down the infra container.  I'm not sure about the rkt runtime.  But depending on the runtime to call TearDownPod() on a SetUpPod() failure seems pretty fragile, and I think the plugins should do it themsevles if they can.

freehan replied:
Agreed. That is mostly because network plugin is event triggered. 

I was thinking adding a  Reconcile interface into network plugin. It can be triggered periodically and do whatever it want to ensure/cleanup network configuration.

dcbw commented:
Still flaking on #26431...

danwinship commented on pkg/kubelet/network/hostport/fake_iptables.go:
```diff
@@ -0,0 +1,335 @@
+	return false, nil
+}
+
+func normalizeRule(rule string) (string, error) {
```
do we have a shell-command-line-parsing utility somewhere?

maybe this function should panic/fail if it hits something that it's not going to parse correctly (eg, `\"` or `\ `)

danwinship replied:
oh, or just fix it to normalize the `[]string` form rather than the `string` form

dcbw replied:
It's not easy to do []string normalization, because the rules coming from Save()/SaveAll() are already in string-form, and we'd need to split them apart into []string *and* handle quoted comments.

dcbw committed:
kubenet: clean up networking when setup errors occur

Relying on the runtime to later call cleanup is fragile, so make sure
that everything gets nicely cleaned up when setup errors occur.

dcbw commented:
@freehan @danwinship PTAL, thanks!

dcbw commented:
+ go install ./cmd/...
Build was aborted
Aborted by Mike Danese
[xUnit] [INFO] - Starting to record.

dcbw commented:
@k8s-bot test this issue: #IGNORE

freehan commented on pkg/kubelet/network/hostport/hostport.go:
```diff
@@ -0,0 +1,385 @@
+	}
+}
+
+func NewTestHostportHandler(iptables utiliptables.Interface, portOpener hostportOpener) HostportHandler {
```
Maybe put this in hostport_test?

freehan commented on pkg/kubelet/network/hostport/hostport.go:
```diff
@@ -0,0 +1,385 @@
+		return retErr
+	}
+
+	for port, socket := range ports {
```
nit: s/port/hostport

freehan commented on pkg/kubelet/network/hostport/hostport.go:
```diff
@@ -0,0 +1,385 @@
+	return nil
+}
+
+// gatherAllHostports returns all hostports that should be presented on node
```
nit: update function comment

dcbw commented:
@freehan @danwinship PTAL, thanks!

freehan commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -467,6 +491,36 @@ func (plugin *kubenetNetworkPlugin) Status() error {
+	runningPods := make([]*hostport.RunningPod, 0)
+	for _, p := range pods {
+		for _, c := range p.Containers {
+			if c.Name != dockertools.PodInfraContainerName {
```
Add a comment here for something like "Assuming pod specs yield by runtime.GetPods include name and ID of containers in pod"

freehan commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -365,61 +327,123 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
+	if err := plugin.setup(namespace, name, id, pod); err != nil {
+		// Make sure everything gets cleaned up on errors
+		podIP, _ := plugin.podIPs[id]
+		plugin.teardown(namespace, name, id, podIP)
```
Maybe log the err from teardown ?

dcbw commented:
@freehan updated for your comments, PTAL

freehan commented:
LGTM. This is great! We should get it in. Maybe for 1.3.1
Pending on tests.
